src/storage/write_ahead_log/log_serializer_task.cpp

```diff
@@ -148,26 +184,44 @@ void LogSerializerTask::HandFilledBufferToWriter() {
   NOISEPAGE_ASSERT(primary_replication_manager_ != DISABLED ||
                        retention_policy != transaction::RetentionPolicy::RETENTION_LOCAL_DISK_AND_NETWORK_REPLICAS,
                    "If replication is disabled, then you can't send buffers to replicas.");
+  if (destination == SerializeDestination::DISK) {
+    if (disk_filled_buffer_ != nullptr) {
+      disk_filled_buffer_->PrepareForSerialization(retention_policy);
+    }
+    // Hand over the filled buffer
+    disk_filled_buffer_queue_->Enqueue(std::make_pair(disk_filled_buffer_, disk_commits_in_buffer_));
+    // Signal disk log consumer task thread that a buffer has been handed over
+    disk_log_writer_thread_cv_->notify_one();
+    // Mark that the task doesn't have a buffer in its possession to which it can write to
+    disk_commits_in_buffer_.clear();
+    disk_filled_buffer_ = nullptr;
+  } else if (destination == SerializeDestination::REPLICAS) {
+    if (replication_filled_buffer_ != nullptr) {
+      replication_filled_buffer_->PrepareForSerialization(retention_policy);
+
+      // Replicate the buffer if it exists.
+      // TODO(WAN): Note that this is effectively on the critical path to commit callbacks being invoked. Aka terrible.
+      if (primary_replication_manager_ != DISABLED) {
```

Comment:
These nested ifs are getting pretty spaghetti-code-like. It's hard to follow the logic. Can the replication manager even be disabled if you're trying to replicate? If not that should be an assertion. I feel like a lot of the permutations in these ifs aren't actually possible, but it's tough to follow the logic with no many nested ifs.

Reply:
@lmwnshn

Reply:
From last last thursday, I think we said we're going to split up `RetentionPolicy` into something like the `TransactionPolicy` draft in #1525 (which does explicitly add an assertion of the above flavor). That would allow us to distinguish between our retention policy for the disk and our retention policy for the replica.

I'm not sure I get your question, sorry. Are you asking me if it is technically impossible?